File: v1/server/src/db/routes/saints.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from src.db.routes.auth import require_admin
from src.schemas.saints import SaintCreate, SaintRead, SaintUpdate
from src.schemas.admin import SaintWithStats
from src.schemas.attendance import CheckInCreate, CheckInResponse
from src.services.saints_service import (
    register_saint,
    update_saint,
    search_saint,
    list_saints,
    get_saint_with_stats,
)
from src.services.checkin_service import check_in_saint
from src.db.setup import get_session
from advanced_alchemy.exceptions import DuplicateKeyError


saints_router = APIRouter(prefix="", tags=["saints"])


@saints_router.post("/register", status_code=status.HTTP_201_CREATED, response_model=SaintRead)
async def register_saint_route(
    data: SaintCreate,
    db: AsyncSession = Depends(get_session),
):
    try:
        return await register_saint(db, data)
    except DuplicateKeyError:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="This email address is already registered.",
        )


@saints_router.get("/search", response_model=SaintRead)
async def search_saint_route(
    first_name: str,
    last_name: str,
    db: AsyncSession = Depends(get_session),
):
    saint = await search_saint(db, first_name, last_name)
    if saint is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Saint not found")
    return saint


@saints_router.get("", response_model=list[SaintRead], dependencies=[Depends(require_admin)])
async def list_saints_route(db: AsyncSession = Depends(get_session)):
    return await list_saints(db)


@saints_router.get("/{saint_id}", response_model=SaintWithStats, dependencies=[Depends(require_admin)])
async def get_saint_route(
    saint_id: uuid.UUID,
    db: AsyncSession = Depends(get_session),
):
    result = await get_saint_with_stats(db, saint_id)
    if result is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Saint not found")
    saint, attendance_count, last_seen = result
    return SaintWithStats.model_validate(
        {**saint.__dict__, "attendance_count": attendance_count, "last_seen": last_seen}
    )


@saints_router.patch("/{saint_id}", response_model=SaintRead, dependencies=[Depends(require_admin)])
async def update_saint_route(
    saint_id: uuid.UUID,
    data: SaintUpdate,
    db: AsyncSession = Depends(get_session),
):
    saint = await update_saint(db, saint_id, data)
    if saint is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Saint not found")
    return saint


@saints_router.post("/checkin", response_model=CheckInResponse)
async def checkin_saint_route(
    data: CheckInCreate,
    db: AsyncSession = Depends(get_session),
):
    return await check_in_saint(db, data.saint_id)

File: v1/client/src/api/saints.ts
import { get, post, patch, ApiError } from './client'
import type { Saint, SaintCreate, SaintWithStats, AdminStats, CheckInResponse, ReportData } from './types'

export async function searchSaint(firstName: string, lastName: string): Promise<Saint | null> {
  try {
    return await get<Saint>('/saints/search', {
      first_name: firstName,
      last_name: lastName,
    })
  } catch (e) {
    if (e instanceof ApiError && e.status === 404) return null
    throw e
  }
}

export async function registerSaint(data: SaintCreate): Promise<Saint> {
  return post<Saint>('/saints/register', data)
}

export async function updateSaint(id: string, data: Partial<SaintCreate>): Promise<Saint> {
  return patch<Saint>(`/saints/${id}`, data)
}

export async function checkInSaint(saintId: string): Promise<CheckInResponse> {
  return post<CheckInResponse>('/saints/checkin', { saint_id: saintId })
}

export async function listSaints(): Promise<Saint[]> {
  return get<Saint[]>('/saints')
}

export async function getSaint(id: string): Promise<SaintWithStats> {
  return get<SaintWithStats>(`/saints/${id}`)
}

export async function getAdminStats(): Promise<AdminStats> {
  return get<AdminStats>('/admin/stats')
}

export async function getReportData(period: 'weekly' | 'monthly' | 'all'): Promise<ReportData> {
  return get<ReportData>('/admin/report', { period })
}
